network/p2p/dns/resolver.go

```diff
@@ -104,18 +173,12 @@ func (r *Resolver) lookupIPAddr(ctx context.Context, domain string) ([]net.IPAdd
 		return r.lookupResolverForIPAddr(ctx, domain)
 	}
 
-	if !fresh && r.shouldResolveIP(domain) {
-		r.unit.Launch(func() {
-			_, err := r.lookupResolverForIPAddr(ctx, domain)
-			if err != nil {
-				// invalidates cached entry when hits error on resolving.
-				invalidated := r.c.invalidateIPCacheEntry(domain)
-				if invalidated {
-					r.collector.OnDNSCacheInvalidated()
-				}
-			}
-			r.doneResolvingIP(domain)
-		})
+	if !fresh && r.shouldResolveIP(domain) && !util.CheckClosed(r.cm.ShutdownSignal()) {
+		select {
+		case r.ipRequests <- &lookupIPRequest{domain}:
+		default:
+			r.logger.Warn().Str("domain", domain).Msg("IP lookup request queue is full, dropping request")
+		}
```

Comment:
🚩  This part is needed to get refactored to make sure updating an expired DNS request is not consigned to starvation, i.e., Each time the DNS cache is queried for a domain that is expired, the DNS cache is out of available workers, and hence keeps only returning that expired entry. This can simply happen if the DNS query rate outpaces the number of workers' resolution rate. The previous approach was taking care of this edge case. You may revert to the previous version (i.e., `engine.Unit`-based one) if resolving this edge case is not trivial.

Reply:
The previous approach takes care of this edge case because it was not worker-based, so in theory there could be an infinite number of go-routines spawned. In a worker-based design, it is impossible to avoid this issue, because you have a finite number of workers and you cannot have a queue of infinite length. Dropping requests when a queue is full is the standard approach in scenarios like this. The alternative solution is to block callers when the queue is full (ie backpressure), but in this scenario I don't think that's a better approach.

Reply:
Correct, then I would suggest reverting it to the previous version that was using goroutines through `unit.Launch`. Running a `goroutine` per _expired_ (and not all) DNS requests is still an efficient approach considering our system scale. The benchmarking results also do not show any `goroutine` overheads, hence the infinite number of spawned `goroutines` is still not a concern.

The current worker-based version poses a starvation concern and such issues would be extremely hard to fix in production. To be correctly functional, we need an intermediate mempool that sits between the resolver and workers. So that DNS requests are written to the mempool, then the resolver takes requests from mempools and spins up workers. Hence, no DNS request is ever facing starvation. This is a missing functionality in the current approach and breaks the liveness of the code, i.e., there is a likelihood that an expired DNS request remains expired for a long while. We have this worker-based DNS cache issued up [here](https://github.com/dapperlabs/flow-go/issues/5868) to be architected and implemented later.

Reply:
I want to avoid using `unit` since this is something being deprecated, so new code should not use it.

For now, using a large-enough channel buffer size and blocking callers in case the channel gets full should be good enough for the current system scale, and this can be easily updated later by replacing the channel with a proper mempool implementation. This way we can do a chunk of the worker-based implementation for https://github.com/dapperlabs/flow-go/issues/5868, and leave the remaining work for later.

Reply:
Or instead of blocking callers, we can add a new metric to track how often the request queue really gets full and requests get dropped, and we can track this on production. My guess is that with a sufficiently large buffer size it will never happen

Reply:
> Or instead of blocking callers, we can add a new metric to track how often the request queue really gets full and requests get dropped, and we can track this on production. My guess is that with a sufficiently large buffer size it will never happen

Please go with this one, and I will expedite https://github.com/dapperlabs/flow-go/issues/5868 to implement the buffer. Also, please make sure that this new metric is scrapped on all of our networks.

Reply:
166aef6